File: GHCAA.Infrastructure/Services/ForumService.cs
using Ganss.Xss;
using GHCAA.Application.DTOs;
using GHCAA.Application.Interfaces;
using GHCAA.Domain.Models;
using GHCAA.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace GHCAA.Infrastructure.Services
{
    public class ForumService : IForumService
    {
        private readonly ApplicationDbContext _context;
        // 1d: Shared, stateless sanitizer instance — HtmlSanitizer is thread-safe.
        private static readonly HtmlSanitizer _sanitizer = new();

        public ForumService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<List<ForumCategoryDto>> GetCategoriesAsync()
        {
            return await _context.ForumCategories
                .Select(c => new ForumCategoryDto
                {
                    Id = c.Id,
                    Name = c.Name,
                    Description = c.Description,
                    SortOrder = c.SortOrder,
                    TopicCount = c.Topics.Count(t => t.IsActive),
                    PostCount = c.Topics.SelectMany(t => t.Posts).Count(p => p.IsActive)
                })
                .OrderBy(c => c.SortOrder)
                .ThenBy(c => c.Name)
                .ToListAsync();
        }

        public async Task<List<ForumTopicDto>> GetTopicsByCategoryAsync(int categoryId, int page, int pageSize)
        {
            var query = _context.ForumTopics
                .Include(t => t.Author)
                .Where(t => t.CategoryId == categoryId);

            return await query
                .OrderByDescending(t => t.IsPinned)
                .ThenByDescending(t => t.LastUpdatedAt ?? t.CreatedAt)
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .Select(t => new ForumTopicDto
                {
                    Id = t.Id,
                    CategoryId = t.CategoryId,
                    Title = t.Title,
                    Content = t.Content,
                    AuthorId = t.AuthorId,
                    AuthorName = t.Author != null ? t.Author.FullName : "Unknown",
                    AuthorPhotoUrl = t.Author != null ? t.Author.PhotoPath : null,
                    CreatedAt = t.CreatedAt,
                    LastUpdatedAt = t.LastUpdatedAt,
                    ViewCount = t.ViewCount,
                    IsPinned = t.IsPinned,
                    IsLocked = t.IsLocked,
                    ReplyCount = t.Posts.Count(p => p.IsActive)
                })
                .ToListAsync();
        }

        public async Task<ForumTopicDto?> GetTopicByIdAsync(int topicId)
        {
            var topic = await _context.ForumTopics
                .Include(t => t.Author)
                .FirstOrDefaultAsync(t => t.Id == topicId);

            if (topic == null) return null;

            // Increment view count
            topic.ViewCount++;
            await _context.SaveChangesAsync();

            return new ForumTopicDto
            {
                Id = topic.Id,
                CategoryId = topic.CategoryId,
                Title = topic.Title,
                Content = topic.Content,
                AuthorId = topic.AuthorId,
                AuthorName = topic.Author != null ? topic.Author.FullName : "Unknown",
                AuthorPhotoUrl = topic.Author != null ? topic.Author.PhotoPath : null,
                CreatedAt = topic.CreatedAt,
                LastUpdatedAt = topic.LastUpdatedAt,
                ViewCount = topic.ViewCount,
                IsPinned = topic.IsPinned,
                IsLocked = topic.IsLocked,
                ReplyCount = await _context.ForumPosts.CountAsync(p => p.TopicId == topicId && p.IsActive)
            };
        }

        public async Task<List<ForumPostDto>> GetPostsByTopicAsync(int topicId, int page, int pageSize)
        {
            return await _context.ForumPosts
                .Include(p => p.Author)
                .Where(p => p.TopicId == topicId)
                .OrderBy(p => p.CreatedAt)
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .Select(p => new ForumPostDto
                {
                    Id = p.Id,
                    TopicId = p.TopicId,
                    Content = p.Content,
                    AuthorId = p.AuthorId,
                    AuthorName = p.Author != null ? p.Author.FullName : "Unknown",
                    AuthorPhotoUrl = p.Author != null ? p.Author.PhotoPath : null,
                    CreatedAt = p.CreatedAt,
                    UpdatedAt = p.UpdatedAt,
                    ParentPostId = p.ParentPostId
                })
                .ToListAsync();
        }

        public async Task<ForumTopicDto> CreateTopicAsync(CreateForumTopicDto dto, int authorId)
        {
            var category = await _context.ForumCategories.FindAsync(dto.CategoryId);
            if (category == null || !category.IsActive)
                throw new Exception("Invalid or inactive category.");

            var topic = new ForumTopic
            {
                CategoryId = dto.CategoryId,
                Title = dto.Title,
                Content = _sanitizer.Sanitize(dto.Content ?? ""), // 1d: strip XSS before storage
                AuthorId = authorId,
                CreatedAt = DateTime.UtcNow,
                IsActive = true
            };

            _context.ForumTopics.Add(topic);
            await _context.SaveChangesAsync();

            return await GetTopicByIdAsync(topic.Id) ?? throw new Exception("Topic created but could not be retrieved.");
        }

        public async Task<ForumPostDto> CreatePostAsync(CreateForumPostDto dto, int authorId)
        {
            var topic = await _context.ForumTopics.FindAsync(dto.TopicId);
            if (topic == null || !topic.IsActive)
                throw new Exception("Topic not found or is inactive.");

            if (topic.IsLocked)
                throw new Exception("Topic is locked and cannot receive new posts.");

            var post = new ForumPost
            {
                TopicId = dto.TopicId,
                Content = _sanitizer.Sanitize(dto.Content ?? ""), // 1d: strip XSS before storage
                AuthorId = authorId,
                ParentPostId = dto.ParentPostId,
                CreatedAt = DateTime.UtcNow,
                IsActive = true
            };

            _context.ForumPosts.Add(post);
            
            topic.LastUpdatedAt = DateTime.UtcNow; // Bump topic

            await _context.SaveChangesAsync();

            return new ForumPostDto
            {
                Id = post.Id,
                TopicId = post.TopicId,
                Content = post.Content,
                AuthorId = post.AuthorId,
                AuthorName = "Self", // Client typically knows who they are, but a reload from DB is safer
                CreatedAt = post.CreatedAt
            };
        }

        public async Task DeleteTopicAsync(int topicId, int memberId, bool isSuperAdmin)
        {
            var topic = await _context.ForumTopics.FindAsync(topicId);
            if (topic == null) return;

            if (topic.AuthorId != memberId && !isSuperAdmin)
                throw new UnauthorizedAccessException("You can only delete your own topics.");

            topic.IsActive = false;
            await _context.SaveChangesAsync();
        }

        public async Task DeletePostAsync(int postId, int memberId, bool isSuperAdmin)
        {
            var post = await _context.ForumPosts.FindAsync(postId);
            if (post == null) return;

            if (post.AuthorId != memberId && !isSuperAdmin)
                throw new UnauthorizedAccessException("You can only delete your own posts.");

            post.IsActive = false;
            await _context.SaveChangesAsync();
        }
    }
}

File: GHCAA.Mobile/lib/features/forum/forum_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../core/api/api_client.dart';

class ForumCategory {
  final int id;
  final String name;
  final String? description;
  final int sortOrder;
  final int topicCount;
  final int postCount;

  ForumCategory({
    required this.id,
    required this.name,
    this.description,
    required this.sortOrder,
    required this.topicCount,
    required this.postCount,
  });

  factory ForumCategory.fromJson(Map<String, dynamic> json) {
    return ForumCategory(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      description: json['description'],
      sortOrder: json['sortOrder'] ?? 0,
      topicCount: json['topicCount'] ?? 0,
      postCount: json['postCount'] ?? 0,
    );
  }
}

class ForumTopic {
  final int id;
  final int categoryId;
  final String title;
  final String content;
  final int authorId;
  final String authorName;
  final String? authorPhotoUrl;
  final DateTime createdAt;
  final DateTime? lastUpdatedAt;
  final int viewCount;
  final bool isPinned;
  final bool isLocked;
  final int replyCount;

  ForumTopic({
    required this.id,
    required this.categoryId,
    required this.title,
    required this.content,
    required this.authorId,
    required this.authorName,
    this.authorPhotoUrl,
    required this.createdAt,
    this.lastUpdatedAt,
    required this.viewCount,
    required this.isPinned,
    required this.isLocked,
    required this.replyCount,
  });

  factory ForumTopic.fromJson(Map<String, dynamic> json) {
    return ForumTopic(
      id: json['id'] ?? 0,
      categoryId: json['categoryId'] ?? 0,
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      authorId: json['authorId'] ?? 0,
      authorName: json['authorName'] ?? '',
      authorPhotoUrl: json['authorPhotoUrl'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
      lastUpdatedAt: json['lastUpdatedAt'] != null ? DateTime.parse(json['lastUpdatedAt']) : null,
      viewCount: json['viewCount'] ?? 0,
      isPinned: json['isPinned'] ?? false,
      isLocked: json['isLocked'] ?? false,
      replyCount: json['replyCount'] ?? 0,
    );
  }
}

class ForumPost {
  final int id;
  final int topicId;
  final String content;
  final int authorId;
  final String authorName;
  final String? authorPhotoUrl;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final int? parentPostId;

  ForumPost({
    required this.id,
    required this.topicId,
    required this.content,
    required this.authorId,
    required this.authorName,
    this.authorPhotoUrl,
    required this.createdAt,
    this.updatedAt,
    this.parentPostId,
  });

  factory ForumPost.fromJson(Map<String, dynamic> json) {
    return ForumPost(
      id: json['id'] ?? 0,
      topicId: json['topicId'] ?? 0,
      content: json['content'] ?? '',
      authorId: json['authorId'] ?? 0,
      authorName: json['authorName'] ?? '',
      authorPhotoUrl: json['authorPhotoUrl'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
      parentPostId: json['parentPostId'],
    );
  }
}

class ForumService {
  final Dio _dio;

  ForumService(this._dio);

  Future<List<ForumCategory>> getCategories() async {
    try {
      final response = await _dio.get('/forum/categories');
      if (response.statusCode == 200) {
        return (response.data as List).map((c) => ForumCategory.fromJson(c)).toList();
      }
    } catch (e) {
      debugPrint('ForumService.getCategories failed: $e');
      rethrow;
    }
    return [];
  }

  Future<List<ForumTopic>> getTopics(int categoryId, {int page = 1, int pageSize = 20}) async {
    try {
      final response = await _dio.get(
        '/forum/categories/$categoryId/topics',
        queryParameters: {'page': page, 'pageSize': pageSize},
      );
      if (response.statusCode == 200) {
        return (response.data as List).map((t) => ForumTopic.fromJson(t)).toList();
      }
    } catch (e) {
      debugPrint('ForumService.getTopics failed: $e');
      rethrow;
    }
    return [];
  }

  Future<ForumTopic?> getTopic(int topicId) async {
    try {
      final response = await _dio.get('/forum/topics/$topicId');
      if (response.statusCode == 200) {
        return ForumTopic.fromJson(response.data);
      }
    } catch (e) {
      debugPrint('ForumService.getTopic failed: $e');
    }
    return null;
  }

  Future<List<ForumPost>> getPosts(int topicId, {int page = 1, int pageSize = 100}) async {
    try {
      final response = await _dio.get(
        '/forum/topics/$topicId/posts',
        queryParameters: {'page': page, 'pageSize': pageSize},
      );
      if (response.statusCode == 200) {
        return (response.data as List).map((p) => ForumPost.fromJson(p)).toList();
      }
    } catch (e) {
      debugPrint('ForumService.getPosts failed: $e');
      rethrow;
    }
    return [];
  }

  Future<ForumTopic?> createTopic(int categoryId, String title, String content) async {
    try {
      final response = await _dio.post(
        '/forum/topics',
        data: {
          'categoryId': categoryId,
          'title': title,
          'content': content,
        },
      );
      if (response.statusCode == 201 || response.statusCode == 200) {
        return ForumTopic.fromJson(response.data);
      }
    } catch (e) {
      debugPrint('ForumService.createTopic failed: $e');
    }
    return null;
  }

  Future<ForumPost?> createPost(int topicId, String content, {int? parentPostId}) async {
    try {
      final response = await _dio.post(
        '/forum/topics/$topicId/posts',
        data: {
          'topicId': topicId,
          'content': content,
          'parentPostId': parentPostId,
        },
      );
      if (response.statusCode == 200 || response.statusCode == 201) {
        return ForumPost.fromJson(response.data);
      }
    } catch (e) {
      debugPrint('ForumService.createPost failed: $e');
    }
    return null;
  }

  Future<bool> deleteTopic(int topicId) async {
    try {
      final response = await _dio.delete('/forum/topics/$topicId');
      return response.statusCode == 204 || response.statusCode == 200;
    } catch (e) {
      debugPrint('ForumService.deleteTopic failed: $e');
      return false;
    }
  }

  Future<bool> deletePost(int postId) async {
    try {
      final response = await _dio.delete('/forum/posts/$postId');
      return response.statusCode == 204 || response.statusCode == 200;
    } catch (e) {
      debugPrint('ForumService.deletePost failed: $e');
      return false;
    }
  }
}

final forumServiceProvider = Provider<ForumService>((ref) {
  return ForumService(ref.read(dioProvider));
});

final forumCategoriesProvider = FutureProvider.autoDispose<List<ForumCategory>>((ref) async {
  return ref.read(forumServiceProvider).getCategories();
});

final forumTopicsProvider = FutureProvider.family.autoDispose<List<ForumTopic>, int>((ref, categoryId) async {
  return ref.read(forumServiceProvider).getTopics(categoryId);
});

final topicDetailProvider = FutureProvider.family.autoDispose<ForumTopic?, int>((ref, topicId) async {
  return ref.read(forumServiceProvider).getTopic(topicId);
});

final topicPostsProvider = FutureProvider.family.autoDispose<List<ForumPost>, int>((ref, topicId) async {
  return ref.read(forumServiceProvider).getPosts(topicId);
});
